tfhe/src/integer/block_decomposition.rs

```diff
@@ -159,11 +194,18 @@ where
 
         if self.num_bits_valid < self.num_bits_in_mask {
             // This will be the case when self.num_bits_in_mask is not a multiple
-            // of T::BITS.  We replace bits that
-            // do not come from the actual T but from the padding
-            // intoduced by the shift, to a specific value.
-            for i in self.num_bits_valid..self.num_bits_in_mask {
-                masked |= self.padding_bit << i;
+            // of T::BITS.
+            //
+            // We replace bits that do not come from the actual T but from the padding
+            // intoduced by the shift, to a specific value. if one was provided.
+            if let Some(padding_bit) = self.padding_bit {
+                let padding_mask = (self.bit_mask >> self.num_bits_valid) << self.num_bits_valid;
+                masked = masked & !padding_mask;
+
+                let padding_bit = T::cast_from(padding_bit);
+                for i in self.num_bits_valid..self.num_bits_in_mask {
+                    masked |= padding_bit << i;
```

Comment:
didn't you say this does not work if padding_bit is 0 ? Or am I misreading something ?